compiler/p/codegen/OMRMemoryReference.cpp

```diff
@@ -1316,11 +1317,11 @@ TR::Instruction *OMR::Power::MemoryReference::expandInstruction(TR::Instruction
          if (displacement < LOWER_IMMED || displacement > UPPER_IMMED)
             {
             prevInstruction = generateTrg1ImmInstruction(cg, TR::InstOpCode::lis, node, index, (int16_t)(displacement >> 16), prevInstruction);
-            prevInstruction = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ori, node, index, index, displacement & 0xffff, prevInstruction);
+            prevInstruction = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ori, node, index, index, (uint16_t)(displacement & 0xffff), prevInstruction);
             }
          else if (displacement != 0)
             {
-            prevInstruction = generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, index, displacement, prevInstruction);
+            prevInstruction = generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, index, (int16_t)displacement, prevInstruction);
```

Comment:
We should try and avoid doing C style casts as it is not type safe. Try to prefer C++ style `static_cast` since the compiler will verify we are not truncating the result in any way which are all potential silent bugs, especially with signed/unsigned conversions.

Reply:
In these cases we do need to truncate and zero/sign extend.

Reply:
I agree that `static_cast`s are highly preferred. With c-style casts, it's not always clear what the intended behaviour is and that can lead to silent bugs. So, it's generally best to always prefer C++ style casts for both correctness and consistency.

Also, signed/unsigned casts can be particularly tricky because the exact behaviour is not always well specified.

Reply:
Please see my comment above. The truncation and zero/sign extension is intended since the 16-bit value is often obtained by right shifting 32-bit or 64-bit value. The semantics of shifting of signed/unsigned values is even less clear. However it is usually obvious which immediate value is expected by the instruction being generated: signed or unsigned 16 bits. As @aviansie-ben pointed out below it's sometimes redundant but it is safe.

Reply:
Truncation and sign/zero extension are not the issue here.

C-style casts don't make the intent of code clear and are very error-prone. We should always prefer C++ style casts because they are more explicit and easier to reason about.

The code here is already obscure because there are two casts: the explicit cast to a 16-bit integer and the *implicit* cast back to a 32-bit integer (because that's the type of the function parameter). A C++ style cast makes things slightly clearer because it's more explicitly, but there's still a lot of implicit behaviour here. I would actually prefer having a separate API that explicitly states the intention here but I haven't thought too much about what that would like so I won't push for it right now.

Reply:
Sure, I opened a separate issue for that since there are a few places where we still use C style: https://github.com/eclipse/omr/issues/5322